---
title: "AgingGene_Enrichment"
author: "Qi Yan"
date: "07/28/2021"
output: 
  html_notebook:
    toc: TRUE # table of content true
    toc_depth: 3  # upto three depths of headings (specified by #, ## and ###)
    toc_float: true
    number_sections: FALSE
    theme: united  
    highlight: tango
    df_print: paged
    code_folding: show
---

<style type="text/css">

.main-container {
  max-width: 2200px;
  margin-left: auto;
  margin-right: auto;
}
body{ /* Normal  */
      font-size: 16px;
  }
h1.title {
  color: DarkRed;
}
h1 { /* Header 1 */
  color: DarkBlue;
}
h2 { /* Header 2 */
  color: DarkBlue;
}
h3 { /* Header 3 */
  color: DarkBlue;
}
  
</style>
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(anRichment)
library(xlsx)
library(tidyverse)
library(WGCNA)
library(ggpubr)

options(stringsAsFactors = FALSE)
allowWGCNAThreads()
```

https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/GeneAnnotation/

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
GOcollection = buildGOcollection(organism = "human")
```

## Summary of the reference database

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
main_reference <- read.table(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Reference_Database/Aging_Gene_Database_main.tsv", sep = "\t")
colnames(main_reference) <- main_reference[1,]
main_reference <- main_reference[-1,]

main_reference <- 
  main_reference %>%
  dplyr::mutate(study = gsub("^([^_]+(?:_[^_]+){3}).*","\\1", Reference))

# Total number of records
print(paste("Total number of records: ", nrow(main_reference), sep = ""))

# Number of studies
print(paste("Number of studies: ", length(unique(main_reference$study)), sep = ""))
print(unique(main_reference$study))

# Number of unique species
print(paste("Number of unique species: ", length(unique(main_reference$Organism)), sep = ""))
print(unique(main_reference$Organism))

# Total number of tables
print(paste("Total number of gene sets (tables): ", length(unique(main_reference$Reference)), sep = ""))

# Number of unique genes
print(paste("Number of unique genes: ", length(unique(main_reference$`Gene Symbol`)), sep = ""))

# Number of unique cell types
print(paste("Number of unique cell types: ", length(unique(main_reference$Cell)), sep = " "))

# organism * study
statistics <- 
  main_reference %>%
  dplyr::select(Organism, study) %>%
  distinct(Organism, study, .keep_all = T) %>%
  group_by(Organism) %>%
  summarise(n=n())

labs <- paste0(statistics$Organism, " (", round(statistics$n/sum(statistics$n)*100,1), "%)")
ggpubr::ggpie(statistics, "n", label = labs,
              lab.pos = "in", lab.font = "white",
              fill = "Organism", color = "white", legend = "right", title = "Percentage of Species")

# Type * study
statistics <- 
  main_reference %>%
  dplyr::select(Type, study) %>%
  distinct(Type, study, .keep_all = T) %>%
  group_by(Type) %>%
  summarise(n=n())

labs <- paste0(statistics$Type, " (", round(statistics$n/sum(statistics$n)*100,1), "%)")
ggpubr::ggpie(statistics, "n", label = labs,
              lab.pos = "in", lab.font = "white",
              fill = "Type", color = "white", legend = "right", title = "Percentage of Type of RNA-seq")

# Genes per study and tissue
statistics <- 
  main_reference %>%
  group_by(study, Tissue) %>%
  summarise(n = n())

ggplot(statistics, aes(fill=Tissue, y=n, x=study)) + 
  geom_bar(position="stack", stat="identity") + 
  ggtitle("Genes per study and tissue") +
  theme(axis.text.x = element_text(angle = 315), axis.text=element_text(size=12),
        axis.title=element_text(size=14,face="bold"))

# Genes per gene set/table
statistics <- 
  main_reference %>%
  group_by(Reference) %>%
  summarise(n = n())

ggplot(statistics, aes(x=Reference, y=n)) +
  geom_segment(aes(x=Reference, xend=Reference, y=0, yend=n), color="skyblue") +
  geom_point( color="blue", size=4, alpha=0.6) +
  ggtitle("Genes per gene set/table") +
  theme_light() +
  coord_flip() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  ) +
  scale_y_continuous(breaks=seq(0,4500,100))
```

## Ortholog Genes

__STEP 1. Need to get an Orthology tables between different species. Got orthology data from various sources__ 

* 1) Based on https://rdrr.io/bioc/annotationTools/man/getHOMOLOG.html, got homologene.data provided by www.ncbi.nlm.nih.gov/HomoloGene, also, orthologs defined in NCBI's 'Orthologs from Annotation pipeline' database (available at ftp.ncbi.nlm.nih.gov/gene/DATA/gene_orthologs.gz, and hence referred to as 'gene_orthologs'). There's also a R package: https://github.com/oganm/homologene. 

* 2) Based on 10.1101/gr.240093.118, and https://github.com/BenayounLaboratory/Mouse_Aging_Epigenomics_2018/tree/master/Figure5_Conservation/Pathway_enrichment/Orthology, I downloaded ortholog maps between mouse, rat, killifish, and human. 

* 3) Also based on Fushan et al. Data S6, I got groups of orthologous sequences assayed in the study.

* 4) From Adriana's file

__STEP 2. First generate files for gene ID conversion__

https://www.biostars.org/p/49606/, https://www.biostars.org/p/446641/ï¼Œ https://yiweiniu.github.io/blog/2018/09/Biological-ID-Conversion/, https://www.bioconductor.org/help/course-materials/2019/BSS2019/05_Annotations.html#biomart

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
################
# Use biomaRt, however it relies on ensembl which doesn't have all the entrez ID information
################

library("biomaRt")

mart = useMart('ensembl')
listDatasets(mart)

# hsapiens_gene_ensembl, mmusculus_gene_ensembl, nfurzeri_gene_ensembl, mfascicularis_gene_ensembl, rnorvegicus_gene_ensembl

# Human
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
listAttributes(human)
genes <- getBM(attributes=c("external_gene_name", "hgnc_symbol","entrezgene_id", "description", "chromosome_name", "start_position", "end_position", "strand", "version", "ensembl_gene_id", "hgnc_id"), mart = human)
write.csv(genes, file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/hsapiens.csv", col.names = T, row.names = F)

# Mouse
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
listAttributes(mouse)
genes <- getBM(attributes=c("external_gene_name", "mgi_symbol","entrezgene_id", "description", "chromosome_name", "start_position", "end_position", "strand", "version", "ensembl_gene_id", "mgi_id"), mart = mouse)
write.csv(genes, file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/mmusculus.csv", col.names = T, row.names = F)

# Killifish
killifish = useMart("ensembl", dataset = "nfurzeri_gene_ensembl")
listAttributes(killifish)
genes <- getBM(attributes=c("external_gene_name","entrezgene_id", "description", "chromosome_name", "start_position", "end_position", "strand", "version", "ensembl_gene_id"), mart = killifish)
write.csv(genes, file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/nfurzeri.csv", col.names = T, row.names = F)

# Macaque
macaque = useMart("ensembl", dataset = "mfascicularis_gene_ensembl")
listAttributes(macaque)
genes <- getBM(attributes=c("external_gene_name","entrezgene_id", "description", "chromosome_name", "start_position", "end_position", "strand", "version", "ensembl_gene_id"), mart = macaque)
write.csv(genes, file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/mfascicularis.csv", col.names = T, row.names = F)

# Rat
rat = useMart("ensembl", dataset = "rnorvegicus_gene_ensembl")
listAttributes(rat)
genes <- getBM(attributes=c("external_gene_name","entrezgene_id", "description", "chromosome_name", "start_position", "end_position", "strand", "version", "ensembl_gene_id"), mart = rat)
write.csv(genes, file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/rnorvegicus.csv", col.names = T, row.names = F)

################
# Use AnnotationHub to get the rest of the entrez ID or symbol
################

# https://seandavi.github.io/ITR/AnnotationHub.html#non-model_organism_gene_annotations

library(AnnotationHub)
query(AnnotationHub(), "^org\\.")
ah <- AnnotationHub()

# Macaca fascicularis
sub_ah = query(ah, c("OrgDb", "Macaca fascicularis"))
sub_ah
orgdb <- query(sub_ah, "OrgDb")[[1]]
columns(orgdb)
keytypes(orgdb) # Type of variables than can be used as keys

original <- read.csv(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/mfascicularis.csv", header = T) # read the csv file which contains missing entrez ID and entrez symbol
symbol <- 
  original %>%
  dplyr::select(external_gene_name) %>%
  dplyr::filter(!is.na(external_gene_name)) %>%
  unique() # get all the gene symbol listed in the csv file and search the orgdb
cols <- c("SYMBOL", "ENTREZID")
map <- 
  select(orgdb, keys=symbol$external_gene_name, columns=cols, keytype="SYMBOL") %>%
  dplyr::filter(!is.na(ENTREZID)) # link it back to the csv file
original <- 
  original %>%
  dplyr::left_join(map, by = c("external_gene_name" = "SYMBOL")) %>%
  dplyr::mutate(entrezgene_id = ifelse(is.na(entrezgene_id), ENTREZID, entrezgene_id)) %>%
  dplyr::select(!ENTREZID)
id <- 
  original %>%
  dplyr::select(entrezgene_id) %>%
  dplyr::filter(!is.na(entrezgene_id)) %>%
  unique() # get all the gene id listed in the csv file and search the orgdb
cols <- c("SYMBOL", "ENTREZID")
map <- 
  select(orgdb, keys=id$entrezgene_id, columns=cols, keytype="ENTREZID") %>%
  dplyr::filter(!is.na(SYMBOL)) # link it back to the csv file
original <- 
  original %>%
  dplyr::left_join(map, by = c("entrezgene_id" = "ENTREZID")) %>%
  dplyr::mutate(external_gene_name = ifelse(external_gene_name == "", SYMBOL, external_gene_name)) %>%
  dplyr::select(!SYMBOL)
write.csv(original, file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/mfascicularis.csv", col.names = T, row.names = F) # save it back

# Nothobranchius furzeri
sub_ah = query(ah, c("OrgDb", "Nothobranchius furzeri"))
sub_ah
orgdb <- query(sub_ah, "OrgDb")[[1]]
columns(orgdb)
keytypes(orgdb) # Type of variables than can be used as keys

original <- read.csv(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/nfurzeri.csv", header = T) # read the csv file which contains missing entrez ID and entrez symbol
symbol <- 
  original %>%
  dplyr::select(external_gene_name) %>%
  dplyr::filter(!is.na(external_gene_name)) %>%
  unique() # get all the gene symbol listed in the csv file and search the orgdb
cols <- c("SYMBOL", "ENTREZID")
map <- 
  select(orgdb, keys=symbol$external_gene_name, columns=cols, keytype="SYMBOL") %>%
  dplyr::filter(!is.na(ENTREZID)) # link it back to the csv file
original <- 
  original %>%
  dplyr::left_join(map, by = c("external_gene_name" = "SYMBOL")) %>%
  dplyr::mutate(entrezgene_id = ifelse(is.na(entrezgene_id), ENTREZID, entrezgene_id)) %>%
  dplyr::select(!ENTREZID)
id <- 
  original %>%
  dplyr::select(entrezgene_id) %>%
  dplyr::filter(!is.na(entrezgene_id)) %>%
  unique() # get all the gene id listed in the csv file and search the orgdb
cols <- c("SYMBOL", "ENTREZID")
map <- 
  select(orgdb, keys=id$entrezgene_id, columns=cols, keytype="ENTREZID") %>%
  dplyr::filter(!is.na(SYMBOL)) # link it back to the csv file
original <- 
  original %>%
  dplyr::left_join(map, by = c("entrezgene_id" = "ENTREZID")) %>%
  dplyr::mutate(external_gene_name = ifelse(external_gene_name == "", SYMBOL, external_gene_name)) %>%
  dplyr::select(!SYMBOL)
write.csv(original, file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/nfurzeri.csv", col.names = T, row.names = F) # save it back

# Rattus Norvegicus
sub_ah = query(ah, c("OrgDb", "Rattus Norvegicus"))
sub_ah
orgdb <- query(sub_ah, "OrgDb")[[1]]
columns(orgdb)
keytypes(orgdb) # Type of variables than can be used as keys

original <- read.csv(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/rnorvegicus.csv", header = T) # read the csv file which contains missing entrez ID and entrez symbol
symbol <- 
  original %>%
  dplyr::select(external_gene_name) %>%
  dplyr::filter(!is.na(external_gene_name)) %>%
  unique() # get all the gene symbol listed in the csv file and search the orgdb
cols <- c("SYMBOL", "ENTREZID")
map <- 
  select(orgdb, keys=symbol$external_gene_name, columns=cols, keytype="SYMBOL") %>%
  dplyr::filter(!is.na(ENTREZID)) # link it back to the csv file
original <- 
  original %>%
  dplyr::left_join(map, by = c("external_gene_name" = "SYMBOL")) %>%
  dplyr::mutate(entrezgene_id = ifelse(is.na(entrezgene_id), ENTREZID, entrezgene_id)) %>%
  dplyr::select(!ENTREZID)
id <- 
  original %>%
  dplyr::select(entrezgene_id) %>%
  dplyr::filter(!is.na(entrezgene_id)) %>%
  unique() # get all the gene id listed in the csv file and search the orgdb
cols <- c("SYMBOL", "ENTREZID")
map <- 
  select(orgdb, keys=id$entrezgene_id, columns=cols, keytype="ENTREZID") %>%
  dplyr::filter(!is.na(SYMBOL)) # link it back to the csv file
original <- 
  original %>%
  dplyr::left_join(map, by = c("entrezgene_id" = "ENTREZID")) %>%
  dplyr::mutate(external_gene_name = ifelse(external_gene_name == "", SYMBOL, external_gene_name)) %>%
  dplyr::select(!SYMBOL)
write.csv(original, file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/rnorvegicus.csv", col.names = T, row.names = F) # save it back

# Mus musculus
sub_ah = query(ah, c("OrgDb", "Mus musculus"))
sub_ah
orgdb <- query(sub_ah, "OrgDb")[[1]]
columns(orgdb)
keytypes(orgdb) # Type of variables than can be used as keys

original <- read.csv(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/mmusculus.csv", header = T) # read the csv file which contains missing entrez ID and entrez symbol
symbol <- 
  original %>%
  dplyr::select(external_gene_name) %>%
  dplyr::filter(!is.na(external_gene_name)) %>%
  unique() # get all the gene symbol listed in the csv file and search the orgdb
cols <- c("SYMBOL", "ENTREZID")
map <- 
  select(orgdb, keys=symbol$external_gene_name, columns=cols, keytype="SYMBOL") %>%
  dplyr::filter(!is.na(ENTREZID)) # link it back to the csv file
original <- 
  original %>%
  dplyr::left_join(map, by = c("external_gene_name" = "SYMBOL")) %>%
  dplyr::mutate(entrezgene_id = ifelse(is.na(entrezgene_id), ENTREZID, entrezgene_id)) %>%
  dplyr::select(!ENTREZID)
id <- 
  original %>%
  dplyr::select(entrezgene_id) %>%
  dplyr::filter(!is.na(entrezgene_id)) %>%
  unique() # get all the gene id listed in the csv file and search the orgdb
cols <- c("SYMBOL", "ENTREZID")
map <- 
  select(orgdb, keys=id$entrezgene_id, columns=cols, keytype="ENTREZID") %>%
  dplyr::filter(!is.na(SYMBOL)) # link it back to the csv file
original <- 
  original %>%
  dplyr::left_join(map, by = c("entrezgene_id" = "ENTREZID")) %>%
  dplyr::mutate(external_gene_name = ifelse(external_gene_name == "", SYMBOL, external_gene_name)) %>%
  dplyr::select(!SYMBOL)
write.csv(original, file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/mmusculus.csv", col.names = T, row.names = F) # save it back

# Homo sapiens
sub_ah = query(ah, c("OrgDb", "Homo sapiens"))
sub_ah
orgdb <- query(sub_ah, "OrgDb")[[1]]
columns(orgdb)
keytypes(orgdb) # Type of variables than can be used as keys

original <- read.csv(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/hsapiens.csv", header = T) # read the csv file which contains missing entrez ID and entrez symbol
symbol <- 
  original %>%
  dplyr::select(external_gene_name) %>%
  dplyr::filter(!is.na(external_gene_name)) %>%
  unique() # get all the gene symbol listed in the csv file and search the orgdb
cols <- c("SYMBOL", "ENTREZID")
map <- 
  select(orgdb, keys=symbol$external_gene_name, columns=cols, keytype="SYMBOL") %>%
  dplyr::filter(!is.na(ENTREZID)) # link it back to the csv file
original <- 
  original %>%
  dplyr::left_join(map, by = c("external_gene_name" = "SYMBOL")) %>%
  dplyr::mutate(entrezgene_id = ifelse(is.na(entrezgene_id), ENTREZID, entrezgene_id)) %>%
  dplyr::select(!ENTREZID)
id <- 
  original %>%
  dplyr::select(entrezgene_id) %>%
  dplyr::filter(!is.na(entrezgene_id)) %>%
  unique() # get all the gene id listed in the csv file and search the orgdb
cols <- c("SYMBOL", "ENTREZID")
map <- 
  select(orgdb, keys=id$entrezgene_id, columns=cols, keytype="ENTREZID") %>%
  dplyr::filter(!is.na(SYMBOL)) # link it back to the csv file
original <- 
  original %>%
  dplyr::left_join(map, by = c("entrezgene_id" = "ENTREZID")) %>%
  dplyr::mutate(external_gene_name = ifelse(external_gene_name == "", SYMBOL, external_gene_name)) %>%
  dplyr::select(!SYMBOL)
write.csv(original, file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/hsapiens.csv", col.names = T, row.names = F) # save it back

# x = c("Tll1", "Tlx3") # gene list to lift over
# genesV2 = getLDS(attributes = c("rgd_symbol"), 
#                  filters = "rgd_symbol", 
#                  values = x , 
#                  mart = rat, 
#                  attributesL = c("hgnc_symbol"), 
#                  martL = human, 
#                  uniqueRows=T)
# 
# library(msigdbr)
# m_df = msigdbr(species = "Rattus norvegicus")
# m_df = msigdbr(species = "Macaca fascicularis")
# m_df_sub = m_df[,c(5,8)] # select only the gene ids necessary
# m_df_sub_dis = distinct(m_df_sub) # remove non unqique rows
```

__STEP 3. Organize ortholog gene database__

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
rat <- read.csv(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/rnorvegicus.csv", header = T) %>%
  dplyr::select(external_gene_name, entrezgene_id) %>%
  dplyr::filter(!is.na(entrezgene_id))
mouse <- read.csv(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/mmusculus.csv", header = T) %>%
  dplyr::select(external_gene_name, entrezgene_id) %>%
  dplyr::filter(!is.na(entrezgene_id))
human <- read.csv(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/hsapiens.csv", header = T) %>%
  dplyr::select(external_gene_name, entrezgene_id) %>%
  dplyr::filter(!is.na(entrezgene_id))
macaque <- read.csv(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/mfascicularis.csv", header = T) %>%
  dplyr::select(external_gene_name, entrezgene_id) %>%
  dplyr::filter(!is.na(entrezgene_id))
killifish <- read.csv(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/geneID_to_geneSymbol/nfurzeri.csv", header = T) %>%
  dplyr::select(external_gene_name, entrezgene_id) %>%
  dplyr::filter(!is.na(entrezgene_id))

# Import NCBI ortholog file, which contains gene list for 376 taxonomies.
ncbi_ortho <- read.table(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/NCBI_gene_orthologs.txt", sep = "\t", header = T)
# list_of_tax <- as.data.frame(unique(c(ncbi_ortho$tax_id, ncbi_ortho$Other_tax_id)))
# temp <- write.table(list_of_tax, file = "C:/Users/QiYan/Desktop/temp.txt", col.names = T, row.names = F, sep = "\t", quote = F)
# Also import tax ID
tax_id <- read.table(file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/Taxonomy_ID.txt", sep = "\t", header = T) %>%
  dplyr::select(taxid, taxname)
# Link the tax id w/ ncbi_ortho, and only keeps human, mouse, rat, killifish, and crab-eating macaque
targeted_tax <- c("Homo sapiens", "Rattus norvegicus", "Mus musculus", "Macaca fascicularis", "Nothobranchius furzeri")

ncbi_ortho_trimmed <-
  ncbi_ortho %>%
  dplyr::left_join(tax_id, by = c("tax_id" = "taxid")) %>%
  dplyr::mutate(tax_id = taxname) %>%
  dplyr::select(-taxname) %>%
  dplyr::rename(taxname_primary = "tax_id") %>%
  dplyr::left_join(tax_id, by = c("Other_tax_id" = "taxid")) %>%
  dplyr::mutate(Other_tax_id = taxname) %>%
  dplyr::select(-taxname) %>%
  dplyr::rename(taxname_other = "Other_tax_id") %>%
  dplyr::filter(taxname_other %in% targeted_tax)

# Create two maps, one maps all the genes to human genome, and one maps all the genes to mouse genome
ncbi_ortho_human <- 
  ncbi_ortho_trimmed %>%
  dplyr::filter(taxname_primary == "Homo sapiens")
ncbi_ortho_mouse <- 
  ncbi_ortho_trimmed %>%
  dplyr::filter(taxname_primary == "Mus musculus")

# link ortholog maps with gene id/symbol file
ncbi_ortho_human_wide <- 
  ncbi_ortho_human %>%
  spread(taxname_other, Other_GeneID) %>% # long to wide
  dplyr::rename(Homo_sapiens_entrezID = "GeneID")
colnames(ncbi_ortho_human_wide) <- c("taxname_primary", "Homo_sapiens_entrezID", "relationship", "Macaca_fascicularis_entrezID", "Mus_musculus_entrezID", "Nothobranchius_furzeri_entrezID", "Rattus_norvegicus_entrezID")
ncbi_ortho_human_wide <- 
  ncbi_ortho_human_wide %>%
  dplyr::left_join(human, by = c("Homo_sapiens_entrezID" = "entrezgene_id")) %>% # human
  dplyr::rename(Homo_sapiens_symbol = "external_gene_name") %>%
  dplyr::left_join(mouse, by = c("Mus_musculus_entrezID" = "entrezgene_id")) %>% # mouse
  dplyr::rename(Mus_musculus_symbol = "external_gene_name") %>%
  dplyr::left_join(rat, by = c("Rattus_norvegicus_entrezID" = "entrezgene_id")) %>% # rat
  dplyr::rename(Rattus_norvegicus_symbol = "external_gene_name") %>%
  dplyr::left_join(killifish, by = c("Nothobranchius_furzeri_entrezID" = "entrezgene_id")) %>% # killifish
  dplyr::rename(Nothobranchius_furzeri_symbol = "external_gene_name") %>%
  dplyr::left_join(macaque, by = c("Macaca_fascicularis_entrezID" = "entrezgene_id")) %>% # macaque
  dplyr::rename(Macaca_fascicularis_symbol = "external_gene_name")
write.csv(ncbi_ortho_human_wide, file = "/Users/qiyan/Dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Ortholog_Genes/ncbi_ortho_human_wide_final.csv", col.names = T, row.names = F)
```



a) Need a ortholog gene map
b) Need a gene id - gene symbol conversion tool for all species

Check the bias of array, do the permutation test randomly select ~ 200 CpGs

## Miscellaneous

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
# format S tables

## Type I
temp <- xlsx::read.xlsx(file = "C:/Users/QiYan/Desktop/temp3.xlsx", sheetIndex = 1)
temp2 <- 
  temp %>%
  dplyr::rename(Cell = celltype) %>%
  dplyr::rename(Gene.Symbol = gene) %>%
  dplyr::mutate(Reference = paste("Zhang_2021_MF_snT_ST2", Cell, sep = "_")) %>%
  dplyr::mutate(Tissue = "Hippocampus ") %>%
  dplyr::mutate(Technology = "Transcriptomes") %>%
  dplyr::mutate(Organism = "Macaca fascicularis") %>%
  dplyr::mutate(Type = "snRNA") %>%
  dplyr::mutate(Trait = "Chronological age") %>%
  dplyr::select(Gene.Symbol, Reference, Technology, Organism, Tissue, Cell, Type, Trait)

write.csv(temp2, file = "C:/Users/QiYan/Desktop/temp2.csv", row.names = F, col.names = T, quote = F)

colnames(temp2)

## Type II
temp <- xlsx::read.xlsx(file = "C:/Users/QiYan/Desktop/temp3.xlsx", sheetIndex = 1)

temp2 <- as.data.frame(unique(unlist(strsplit(temp$Gene, "/"))))
colnames(temp2) <- "Gene"
write.csv(temp2, file = "C:/Users/QiYan/Desktop/temp2.csv", row.names = F, col.names = T, quote = F)

## Type III
temp <- xlsx::read.xlsx(file = "C:/Users/QiYan/Desktop/temp3.xlsx", sheetIndex = 1)

temp2 <- 
  temp %>%
  dplyr::mutate(Reference = paste("Schaum_2020_MM_T_ST3_cluster", ClusterID, sep = "")) %>%
  dplyr::mutate(Tissue = "Multiple ") %>%
  dplyr::mutate(Technology = "Transcriptomics; Proteomics") %>%
  dplyr::mutate(Organism = "Mus musculus") %>%
  dplyr::mutate(Cell = "Bulk") %>%
  dplyr::mutate(Type = "Bulk; scRNA") %>%
  dplyr::mutate(Trait = "Chronological age; Gene expression dynamics with age") %>%
  dplyr::mutate(Note = "Whole-organism gene expression trajectory clustering") %>%
  dplyr::select(Gene.Symbol, Reference, Technology, Organism, Tissue, Cell, Type, Trait, Note)

write.csv(temp2, file = "C:/Users/QiYan/Desktop/temp2.csv", row.names = F, col.names = T, quote = F)

## Type IV
temp <- xlsx::read.xlsx(file = "C:/Users/QiYan/Desktop/temp3.xlsx", sheetIndex = 1)

temp2 <- 
  temp %>%
  dplyr::select(-c(numSig, SigInOne)) %>%
  tidyr::gather(Cell, P_value, -c(gene.name)) %>%
  mutate(Cell = substr(Cell,1,nchar(Cell)-10)) %>%
  dplyr::filter(P_value < 0.1) %>%
  dplyr::select(-P_value) %>%
  dplyr::mutate(Reference = paste("Angelidis_2019_MM_TP_ST5_", Cell, sep = "")) %>%
  dplyr::rename(Gene.Symbol = gene.name) %>%
  dplyr::mutate(Tissue = "Lung ") %>%
  dplyr::mutate(Technology = "Transcriptomics; Proteomics") %>%
  dplyr::mutate(Organism = "Mus musculus") %>%
  # dplyr::mutate(Cell = "Multiple") %>%
  dplyr::mutate(Type = "scRNA") %>%
  dplyr::mutate(Trait = "Chronological age") %>%
  dplyr::mutate(Note = "Cell type-resolved differential gene expression testing between age groups in the single cell data (FDR < 0.1)") %>%
  dplyr::select(Gene.Symbol, Reference, Technology, Organism, Tissue, Cell, Type, Trait, Note)

write.csv(temp2, file = "C:/Users/QiYan/Desktop/temp2.csv", row.names = F, col.names = T, quote = F)

## Type V
temp <- xlsx::read.xlsx(file = "C:/Users/QiYan/Desktop/temp3.xlsx", sheetIndex = 1)
# temp <- read.table(file = "C:/Users/QiYan/Desktop/temp3.txt", sep = "\t")
# colnames(temp) <- temp[1,]
# temp <- temp[-1,]

temp2 <- 
  temp %>%
  tidyr::gather(Cell, P_value, -c(Gene)) %>%
  dplyr::filter(!is.na(P_value)) %>%
  dplyr::select(-P_value) %>%
  # dplyr::mutate(Cell = gsub(".*:","",Cell)) %>%
  dplyr::mutate(Reference = paste("Ximerakis_2019_MM_scT_ST8_neuronal_", Cell, sep = "")) %>%
  # dplyr::mutate(Tissue = gsub("\\..*","",Cell)) %>%
  dplyr::mutate(Tissue = "Brain") %>%
  dplyr::rename(Gene.Symbol = Gene) %>%
  dplyr::mutate(Technology = "Transcriptomics") %>%
  dplyr::mutate(Organism = "Mus_musculus") %>%
  # dplyr::mutate(Cell = "Multiple") %>%
  dplyr::mutate(Type = "scRNA") %>%
  dplyr::mutate(Trait = "Chronological age") %>%
  dplyr::mutate(Note = "Aging-related genes (FDR < 0.05 and FC > 10%) from 3 major neurotransmitter-expressing neuronal subtypes") %>%
  dplyr::select(Gene.Symbol, Reference, Technology, Organism, Tissue, Cell, Type, Trait, Note)

unique(temp2$Cell)
# nrow(temp2[which(temp2$Cell == "global_aging"),])

write.csv(temp2, file = "C:/Users/QiYan/Desktop/temp2.csv", row.names = F, col.names = T, quote = F)
# write.table(temp2, file = "/Users/qiyan/dropbox/Horvath_Lab/HorvathLabCoreMembers/Qi/Aging_Gene/Reference_and_SummaryStatistics/Reference_Database/TabulaMurisSenisZhang_2021_MM_scT_ST2FACS.txt", row.names = F, col.names = T, quote = F, sep = "\t")

## Type VI
temp <- xlsx::read.xlsx(file = "C:/Users/QiYan/Desktop/temp3.xlsx", sheetIndex = 1)

temp2 <- 
  temp %>%
  dplyr::mutate(gene2 = gsub("\\(.*","",Gene)) %>%
  dplyr::mutate(Reference = "Baumgart_2014_NF_T_ST1_Brain") %>%
  dplyr::mutate(Tissue = "Brain") %>%
  dplyr::mutate(Cell = "Bulk") %>%
  dplyr::rename(Gene.Symbol = Gene) %>%
  dplyr::mutate(Technology = "Transcriptomics") %>%
  dplyr::mutate(Organism = "Nothobranchius_furzeri") %>%
  dplyr::mutate(Type = "Bulk") %>%
  dplyr::mutate(Trait = "Chronological age") %>%
  dplyr::mutate(Note = "DEGs as the intersection of three independent statistical tests in at least one of all pairwise comparisons between the age-groups") %>%
  dplyr::select(Gene.Symbol, Reference, Technology, Organism, Tissue, Cell, Type, Trait, Note)

write.csv(temp2, file = "C:/Users/QiYan/Desktop/temp2.csv", row.names = F, col.names = T, quote = F)

```